sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java

```diff
@@ -0,0 +1,250 @@
+
+        policies.add(
+            new UserAgentPolicy(CoreUtils.getApplicationId(clientOptions, logOptions), CLIENT_NAME, CLIENT_VERSION, configuration));
+        policies.add(new RequestIdPolicy());
+
+        policies.addAll(perCallPolicies);
+        HttpPolicyProviders.addBeforeRetryPolicies(policies);
+
+        policies.add(retryPolicy == null ? new RetryPolicy() : retryPolicy);
+        policies.add(new CookiePolicy());
+        policies.add(new AddDatePolicy());
+
+        policies.addAll(perRetryPolicies);
+        HttpPolicyProviders.addAfterRetryPolicies(policies);
+        HttpLoggingPolicy loggingPolicy = new HttpLoggingPolicy(logOptions);
+
+        // We generally put credential policy between BeforeRetry and AfterRetry policies and put Logging policy in the end.
+        // However since ACR uses the rest endpoints of the service in the credential policy,
+        // we want to be able to use the same pipeline (minus the credential policy) to have uniformity in the policy
+        // pipelines across all ACR endpoints.
+        if (credential == null) {
+            logger.verbose("Credentials are null, enabling anonymous access");
+        }
+
+        ArrayList<HttpPipelinePolicy> credentialPolicies = clone(policies);
+        credentialPolicies.add(loggingPolicy);
+
+        ContainerRegistryTokenService tokenService = new ContainerRegistryTokenService(
+            credential,
+            audience,
+            endpoint,
+            serviceVersion,
+            new HttpPipelineBuilder()
+                .policies(credentialPolicies.toArray(new HttpPipelinePolicy[0]))
+                .httpClient(httpClient)
+                .build(),
+            JacksonAdapter.createDefaultSerializerAdapter());
+
+        ContainerRegistryCredentialsPolicy credentialsPolicy = new ContainerRegistryCredentialsPolicy(tokenService);
+        policies.add(credentialsPolicy);
+
+        policies.add(loggingPolicy);
+        HttpPipeline httpPipeline =
+            new HttpPipelineBuilder()
+                .policies(policies.toArray(new HttpPipelinePolicy[0]))
+                .httpClient(httpClient)
+                .build();
+        return httpPipeline;
+    }
+
+    private static ArrayList<HttpPipelinePolicy> clone(ArrayList<HttpPipelinePolicy> policies) {
+        ArrayList<HttpPipelinePolicy> clonedPolicy = new ArrayList<>();
+        for (HttpPipelinePolicy policy:policies) {
+            clonedPolicy.add(policy);
+        }
+
+        return clonedPolicy;
+    }
+
+    /**
+     * This method computes the digest for the buffer content.
+     * Docker digest is a SHA256 hash of the docker image content and is deterministic based on the image build.
+     * @param buffer The buffer containing the image bytes.
+     * @return SHA-256 digest for the given buffer.
+     */
+    public static String computeDigest(ByteBuffer buffer) {
+        int startPosition = buffer.position();
```

Comment:
Instead of capturing and resetting the position a `readOnly` view can be passed into `MessageDigest`.

https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#asReadOnlyBuffer--

Reply:
That internally creates a new buffer. Considering this can be a huge byte array, I would rather be memory conscious. Are there any downsides to moving the position?